### PR TITLE
[ENH] Add DCNNNetwork

### PR DESCRIPTION
**Pull request by aadya940**
Implement Dilated Convolutional Neural Network based Auto-Encoder.

### CODE REVIEW

File: aeon/networks/_dcnn.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    """
+
+    def __init__(
+        self,
+        latent_space_dim=128,
+        num_layers=4,
+        kernel_size=3,
+        activation="relu",
+        num_filters=None,
+        dilation_rate=None,
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:49]: activation not placed in correct order following the docstring

File: aeon/networks/_dcnn.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        if self.dilation_rate is None:
+            self.dilation_rate = [
+                2**layer_num for layer_num in range(1, self.model_depth + 1)
+            ]
+        else:
+            assert isinstance(self.dilation_rate, list)
+            assert len(self.dilation_rate) == self.model_depth
+
+    def build_network(self, input_shape):
+        """Construct a network and return its input and output layers.
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:72]: its preferable to keep the constructor to assigning private members, and all this stuff to be inside the build network function

File: aeon/networks/_dcnn.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        input_shape : tuple of shape = (n_timepoints (m), n_channels (d))
+            The shape of the data fed into the input layer.
+
+        Returns
+        -------
+        encoder : a keras Model.
+        """
+        import tensorflow as tf
+
+        def DCNNLayer(inputs, num_filters, dilation_rate, activation):
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:85]: the function should be a private function and direct member of the class, not a function inside the build network. Should be called _dcnn_layer

File: aeon/networks/_dcnn.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      volume={32},
+      year={2019}
+    }
+    """
+
+    def __init__(
+        self,
+        latent_space_dim=128,
+        num_layers=4,
+        kernel_size=3,
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:46]: i think its good to even give user ability to control the kernel size for each layer, same for activation as well, shouldnt be obligation that its the same through all the network

File: aeon/networks/_dcnn.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+        self.latent_space_dim = latent_space_dim
+        self.kernel_size = kernel_size
+        self.num_filters = num_filters
+        self.model_depth = num_layers
+        self.dilation_rate = dilation_rate
+        self.activation = activation
+
+        if self.num_filters is None:
+            self.num_filters = [32 * i for i in range(1, self.model_depth + 1)]
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:61]: should check the same for num_filters if it was a list, that it has the same length as model_depth

File: aeon/networks/_dcnn.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        activation="relu",
+        num_filters=None,
+        dilation_rate=None,
+    ):
+        super().__init__()
+
+        self.latent_space_dim = latent_space_dim
+        self.kernel_size = kernel_size
+        self.num_filters = num_filters
+        self.model_depth = num_layers
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:56]: naming shouldnt change, should stay num_layers

File: aeon/networks/_dcnn.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        """Construct a network and return its input and output layers.
+
+        Arguments
+        ---------
+        input_shape : tuple of shape = (n_timepoints (m), n_channels (d))
+            The shape of the data fed into the input layer.
+
+        Returns
+        -------
+        encoder : a keras Model.
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:81]: the build_network at the end should return model not encoder (check first comment above)

File: aeon/networks/_dcnn.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
+"""Encoder using Dilated Convolutional Nerual Networks (DCNN)."""
+
+__maintainer__ = []
+
+from aeon.networks.base import BaseDeepNetwork
+
+
+class DCNNEncoderNetwork(BaseDeepNetwork):
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:8]: If am not mistaken DCNN is not an auto encoder based, so you can actually remove the naming "Encoder" and just keep DCNNNetwork

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
good start, a few comments on the network

**Comment by aadya940** [comment]
Done :-) , Thanks for the quick review.

**Comment by hadifawaz1999** [comment]
make an issue to add again the weight normalization code

File: aeon/networks/_dcnn.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
+"""Dilated Convolutional Nerual Networks (DCNN) Model."""
+
+__maintainer__ = []
+
+from aeon.networks.base import BaseDeepNetwork
+
+
+class DCNNNetwork(BaseDeepNetwork):
+    """Establish the network structure for a DCNN-Model.
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:9]: do not override private members

**Review by hadifawaz1999 — DISMISSED** [review]
lgtm!

**Comment by MatthewMiddlehurst** [comment]
> make an issue to add again the weight normalization code

Was this done?

**Comment by aadya940** [comment]
> > make an issue to add again the weight normalization code
> 
> Was this done?

Nope, will do it.

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
LGTM this one given its not an AUTO ENCODER network, you can remove the temporal_latent_space thing
And you should add the _config parameter

File: aeon/networks/tests/test_dcnn.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
+"""Tests for the DCNN Model."""
+
```
Comment (hadifawaz1999) [inline aeon/networks/tests/test_dcnn.py:2]: temporal_latent_space should be removed from the tests also for this network

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]

**Comment by hadifawaz1999** [comment]
there is a conflict with main @aadya940

File: aeon/networks/_dcnn.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+    Dilated Convolutional Neural Network based Model
+    for low-rank embeddings.
+
+    Parameters
+    ----------
+    latent_space_dim: int, default=128
+        Dimension of the models's latent space.
+    num_layers: int, default=4
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:18]: i would prefer if n_layers is the same on all networks instead of num_layers

File: aeon/networks/_dcnn.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    ----------
+    latent_space_dim: int, default=128
+        Dimension of the models's latent space.
+    n_layers: int, default=4
+        Number of convolution layers.
+    kernel_size: int, default=3
+        Size of the 1D Convolutional Kernel.
+    activation: str, default="relu"
+        The activation function used by convolution layers.
+    num_filters: int, default=None
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:24]: n_filters

File: aeon/networks/_dcnn.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        "structure": "encoder",
+    }
+
+    def __init__(
+        self,
+        latent_space_dim=128,
+        n_layers=4,
+        kernel_size=3,
+        activation="relu",
+        num_filters=None,
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:54]: n_filters

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]

File: aeon/networks/_dcnn.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+        The activation function used by convolution layers.
+    n_filters: int, default=None
+        Number of filters used in convolution layers.
+    dilation_rate: list, default=None
+        The dilation rate for convolution.
+
+    References
+    ----------
+    .. [1] Network originally defined in:
+    @article{franceschi2019unsupervised,
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:32]: can you put the citation in non bib format ?

File: aeon/networks/_dcnn.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+        self.kernel_size = kernel_size
+        self.n_filters = n_filters
+        self.n_layers = n_layers
+        self.dilation_rate = dilation_rate
+        self.activation = activation
+
+    def build_network(self, input_shape):
+        """Construct a network and return its input and output layers.
+
+        Arguments
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:69]: Parameters instead of Arguments

File: aeon/networks/_dcnn.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+            self._kernel_size = [self.kernel_size for _ in range(self.n_layers)]
+        elif isinstance(self.kernel_size, list):
+            self._kernel_size = self.kernel_size
+            assert len(self.kernel_size) == self.n_layers
+
+        if isinstance(self.activation, str):
+            self._activation = [self.activation for _ in range(self.n_layers)]
+        elif isinstance(self.activation, list):
+            self._activation = self.activation
+            assert len(self._activation) == self.n_layers
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:107]: typo activation instead of _activation in assert

File: aeon/networks/_dcnn.py [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+class DCNNNetwork(BaseDeepLearningNetwork):
+    """Establish the network structure for a DCNN-Model.
+
+    Dilated Convolutional Neural Network based Model
+    for low-rank embeddings.
+
+    Parameters
+    ----------
+    latent_space_dim: int, default=128
+        Dimension of the models's latent space.
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:17]: can you add for the input parameters that are list format what would be the default list in the docstring ?

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
last comments :)

**Comment by aadya940** [comment]
#1962

File: aeon/networks/_dcnn.py (hunk 16 above)
Comment (aadya940) [reply aeon/networks/_dcnn.py:107]: Once, `self._activation` is equal to `self.activation` in line 106, It doesn't really make a difference.

File: aeon/networks/_dcnn.py [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+    n_layers: int, default=4
+        Number of convolution layers.
+    kernel_size: Union[int, List[int]], default=3
+        Size of the 1D Convolutional Kernel. Defaults
+        to a list of three's for `n_layers` elements.
+    activation: Union[str, List[str]], default="relu"
+        The activation function used by convolution layers.
+        Defaults to a list of "relu" for `n_layers` elements.
+    n_filters: Union[int, List[int]], default=None
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:26]: can you add a sentence to n_filters and dilation_rate to how the default values are calculated ?

File: aeon/networks/_dcnn.py [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+        "structure": "encoder",
+    }
+
+    def __init__(
+        self,
+        latent_space_dim=128,
+        n_layers=4,
+        kernel_size=3,
+        activation="relu",
+        n_filters=None,
```
Comment (hadifawaz1999) [inline aeon/networks/_dcnn.py:52]: should have a padding parameter, default to the value used in the original paper, also will be a list for all layers like you handle activation

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
final changes i think, also can you make the activation detached as its own layer instead then inside the conv ?

File: aeon/networks/_dcnn.py (hunk 18 above)
Comment (aadya940) [reply aeon/networks/_dcnn.py:26]: Haven't I already? Check Line 31 and Line 28.

**Review by hadifawaz1999 — APPROVED** [review]

**Pull request merged**